pkg/run/channel_closer.go

```diff
@@ -0,0 +1,116 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package run
+
+import (
+	"context"
+	"sync"
+)
+
+var dummyChannelCloserChan <-chan struct{}
+
+// ChannelCloser can close a goroutine then wait for it to stop.
+type ChannelCloser struct {
+	ctx           context.Context
+	cancel        context.CancelFunc
+	running       sync.WaitGroup
+	waiting       sync.WaitGroup
```

Comment:
Could you rename them to "sender" and "receiver" since the closer is dedicated to chan?